### DISCUSSION TITLE
Populating custom script variables

### DISCUSSION THREAD

**Discussion by bistraque** [netbox-community/netbox#8542] [Q&A] [answered]
A recent use case I have stumbled upon: say i would like to design a custom script to edit some parts of a device Config Context in a friendly form. Obviously I can do a "set" script where form input data replaces current data, but I can't find a way to query device data from script url (e.g. ...?device_id=666) before the form is displayed in order to populate vars with current data before user edit. 

Tried Script `__init__` override but with no luck.

Anyone has an idea, or is it a new feature request?

**Comment by mtinberg** [comment]
I don't know if this helps and I haven't thought this through or tried it but can you use some of the api endpoints from the client side in the way that the various filter drop-down lists are populated to select and fetch the device info using javascript to load into the form elements, rather than having it on the python server side when the page is generated?


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________
From: bistraque ***@***.***>
Sent: Thursday, February 3, 2022 11:18 AM
To: netbox-community/netbox ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [netbox-community/netbox] Populating custom script variables (Discussion #8542)


A recent use case I have stumbled upon: say i would like to design a custom script to edit some parts of a device Config Context in a friendly form. Obviously I can do a "set" script where form input data replaces current data, but I can't find a way to query device data from script url (e.g. ...?device_id=666) before the form is displayed in order to populate vars with current data before user edit.

Tried Script __init__ override but with no luck.

Anyone has an idea, or is it a new feature request?

—
Reply to this email directly, view it on GitHub<https://github.com/netbox-community/netbox/discussions/8542>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAS7UMZW2V2DU4ACE4XCA33UZK2HNANCNFSM5NPRCONQ>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

**Reply by bistraque** [reply]
I am effectively working from server-side only: i.e. how can I do this with "server script" functionality (and avoid plugin development ;) ? I have no control on the way netbox script form is built and displayed.

If I was developing a netbox api Client it wouldn't be much of a problem, it would just be a question of doing the right API queries before form construction and display.

**Comment by candlerb** [comment]
So to clarify the question: are you trying to get access to the query URL which was used when the script was invoked? e.g.

```
/extras/scripts/test_addr.TestAddr/?foo=bar
```

displays the script form, and you want access to "foo=bar" inside your script when the user presses the "Run Script" button?

In principle I think it could be made to work because I can see the POST goes back to the same URL, including the same query string.  But I think Netbox would have to be extended to capture this and pass it to the script, which I believe these days runs as a background task - hence it's not directly connected to the web request.

**Reply by bistraque** [reply]
You are right, this works for device fields which can be set via {{obj.field}} jinja2 syntax in custom links. Unfortunately, this is not possible for custom fields and for json items embedded in device Config Context.

**Comment by bistraque** [comment, marked as answer]
For those interested, I found a solution by overiding Script `as_form()` method. Here is the logic:
```
def as_form(self, data=None, files=None, initial=None):  
    if data is None: # initial form GET request /extras/scripts/Myscript.Myscript/?device=666  
        device = Device.objects.get(id=str(initial['device']))  
        initial['mycf'] = device.custom_field_data['mycf']  
        # other vars ...             
    form = super().as_form(data, files, initial)  
    return form  
```
This method will be called twice: for GET where `data `is None and `initial `contains query params, and for POST where `data `contains final form data.

**Reply by julianstolp** [reply]
Where do I put this code?

**Reply by bistraque** [reply]
If you take [NewBranchScript](https://docs.netbox.dev/en/stable/customization/custom-scripts/) from netbox documentation as an example, it would logically appear between `class Meta:` and `def run(self, data, commit):` blocks.

**Reply by adparis99** [reply]
Out of curiosity, is there a way to add this function in a sort of if-then way? I have a script that generates device configs based on device and config context data. Previously, I would go to the script and select the desired device from a drop-down. Thanks to your answer I can now use a custom link from the device page, which works perfectly to pre-populate the device drop-down. But now the script _only_ works from the custom link; if I try manually going to the script and selecting a device from the dropdown, I get an error. Obviously this is not that big an issue, and if I can't have my cake and eat it too I'm fine with that, just curious if you might know a work around, as anything I've tried as ended in error.

**Reply by bistraque** [reply]
A probable cause: `as_form()` builds the form _before_ you can interact with it using the url params available in the GET (e.g. `?device=666`). Calling the script directly will call `as_form()` with empty params, therefore there is no prepopulation of fields.
Subsequent interactions with the form happen on browser side, so I see no obvious way to get data from the script (unless some ajax builtin feature exists).
